File: go/doubly_linked_list/doubly_linked_list.go
package doublylinkedlist

import "fmt"

type Node struct {
	value          int64
	next, previous *Node
}

func NewNode(value int64) *Node {
	return &Node{
		value:    value,
		next:     nil,
		previous: nil,
	}
}

type DoublyLinkedList struct {
	size uint64
	head *Node
}

func NewDoublyLinkedList() *DoublyLinkedList {
	return &DoublyLinkedList{
		size: 0,
		head: nil,
	}
}

func (d *DoublyLinkedList) insert(value int64) {
	node := NewNode(value)
	d.size++

	if d.head == nil {
		d.head = node
		return
	}

	tmp := d.head

	for tmp != nil {
		if tmp.next == nil {
			node.previous = tmp
			tmp.next = node
			break
		}

		tmp = tmp.next
	}
}

func (d *DoublyLinkedList) insertAtBeginning(value int64) {
	node := NewNode(value)
	d.size++

	if d.head == nil {
		d.head = node
		return
	}

	node.next = d.head
	d.head.previous = node
	d.head = node
}

func (d *DoublyLinkedList) pop() {
	if d.head == nil {
		return
	}

	d.size--
	tmp := d.head

	for tmp != nil {
		if tmp.next == nil {
			tmp.previous.next = nil
			break
		}

		tmp = tmp.next
	}
}

func (d *DoublyLinkedList) show() {
	tmp := d.head

	for tmp != nil {
		fmt.Println(tmp.value)
		tmp = tmp.next
	}
}

File: rust/src/doubly_linked_list.rs
use std::ptr;

struct Node {
    value: i64,
    next: *mut Node,
    previous: *mut Node,
}

impl Node {
    pub fn new(value: i64) -> Self {
        Self {
            value,
            next: ptr::null_mut(),
            previous: ptr::null_mut(),
        }
    }
}

struct DoublyLinkedList {
    size: u64,
    head: *mut Node,
}

impl DoublyLinkedList {
    pub fn new() -> Self {
        Self {
            size: 0,
            head: ptr::null_mut(),
        }
    }

    pub fn insert(&mut self, value: i64) {
        let node = Box::into_raw(Box::new(Node::new(value)));
        self.size += 1;

        if self.head.is_null() {
            self.head = node;
            return;
        }

        let mut tmp = self.head;

        unsafe {
            while !tmp.is_null() {
                if (*tmp).next.is_null() {
                    (*node).previous = tmp;
                    (*tmp).next = node;
                    break;
                }

                tmp = (*tmp).next;
            }
        }
    }

    pub fn insertAtBeginning(&mut self, value: i64) {
        let node = Box::into_raw(Box::new(Node::new(value)));
        self.size += 1;

        if self.head.is_null() {
            self.head = node;
            return;
        }

        unsafe {
            (*node).next = self.head;
            (*self.head).previous = node;
            self.head = node;
        }
    }

    pub fn pop(&mut self) {
        if self.head.is_null() {
            return;
        }

        self.size -= 1;
        let mut tmp = self.head;

        unsafe {
            while !tmp.is_null() {
                if (*tmp).next.is_null() {
                    (*(*tmp).previous).next = ptr::null_mut();
                    break;
                }

                tmp = (*tmp).next;
            }
        }
    }

    pub fn show(&self) {
        let mut tmp = self.head;

        unsafe {
            while !tmp.is_null() {
                println!("{}", (*tmp).value);
                tmp = (*tmp).next;
            }
        }
    }
}
